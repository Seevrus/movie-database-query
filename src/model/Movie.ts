import { MovieResult } from '../types/search-movies';

export class Movie {
  id?: number;

  title?: string;

  originalTitle?: string;

  posterPath?: string | null;

  isAdult?: boolean;

  genreIds?: number[];

  overview?: string;

  popularity?: number;

  voteAverage?: number;

  voteCount?: number;

  releaseDate?: string;

  originalLanguage?: string;

  video?: boolean;

  backdropPath?: string;

  constructor(movieData: MovieResult) {
    this.id = movieData.id;
    this.title = movieData.title;
    this.originalTitle = movieData.original_title;
    this.posterPath = movieData.poster_path;
    this.isAdult = movieData.adult;
    this.genreIds = movieData.genre_ids;
    this.overview = movieData.overview;
    this.popularity = movieData.popularity;
    this.voteAverage = movieData.vote_average;
    this.voteCount = movieData.vote_count;
    this.releaseDate = movieData.release_date;
    this.originalLanguage = movieData.original_language;
    this.video = movieData.video;
    this.backdropPath = movieData.backdrop_path;
  }
}
