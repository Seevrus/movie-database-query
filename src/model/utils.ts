export function jsonFetch<ResponseT>(
  url: string,
  config: RequestInit = {},
): Promise<ResponseT> {
  return fetch(url, config)
    .then((response) => response.json())
    .then((data) => data as ResponseT);
}
