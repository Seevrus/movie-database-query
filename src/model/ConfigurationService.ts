import { ApiConfigurationResponseT, ApiConfigurationT } from '../types/api-configuration';
import { jsonFetch } from './utils';

export const defaultConfiguration = {
  baseUrl: 'https://image.tmdb.org/t/p/',
  size: 'w185',
};

export class ConfigurationService {
  static async getConfiguration(): Promise<ApiConfigurationT> {
    return jsonFetch<ApiConfigurationResponseT>('https://api.themoviedb.org/3/configuration', {
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_API_TOKEN}`,
      },
    })
      .then(response => ({
        baseUrl: response.images.secure_base_url ?? response.images.base_url,
        size: 'w185',
      }))
      .catch(() => defaultConfiguration);
  }
}