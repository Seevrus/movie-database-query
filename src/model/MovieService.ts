import { SearchMoviesT } from '../types/search-movies';
import { Movie } from './Movie';
import { jsonFetch } from './utils';

export class MovieService {
  static async searchMovies(query: string) {
    return jsonFetch<SearchMoviesT>(`https://api.themoviedb.org/3/search/movie?language=en-US&query=${query}&page=1&include_adult=false`, {
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_API_TOKEN}`,
      },
    }).then((response) => {
      return (response.results ?? []).map(movieResult => new Movie(movieResult));
    });
  }
}
