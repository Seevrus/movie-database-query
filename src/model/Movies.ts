import { SearchMoviesT } from '../types/search-movies';
import { Movie } from './Movie';
import { jsonFetch } from './utils';

export class Movies {
  movies: Movie[];

  constructor() {
    this.movies = [];
  }

  async searchMovies(query: string) {
    return jsonFetch<SearchMoviesT>(`https://api.themoviedb.org/3/search/movie?language=en-US&query=${query}&page=1&include_adult=false`, {
      headers: {
        Authorization: `Bearer ${process.env.REACT_APP_API_TOKEN}`,
      },
    }).then((response) => {
      this.movies = (response.results ?? []).map(movieResult => new Movie(movieResult));
    });
  }
}
