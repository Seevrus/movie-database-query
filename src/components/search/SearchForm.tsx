import { useState } from 'react';
import { Alert, Button, Card, Col, Container, Form, Row } from 'react-bootstrap';
import { Movie } from '../../model/Movie';
import { MovieService } from '../../model/MovieService';

interface SearchFormProps {
  setMovies: React.Dispatch<React.SetStateAction<Movie[]>>;
}

const SearchForm = ({ setMovies }: SearchFormProps) => {
  const feedback = false;
  const isFormValidated = false;
  const queryError = '';
  const searchError = '';

  const [queryText, setQueryText] = useState<string>('');

  const onSearch = async () => {
    const movies = await MovieService.searchMovies(queryText);
    setMovies(movies);
  };

  const onQueryChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const taxNumberInput = e.target.value;
    setQueryText(taxNumberInput);
  };

  return (
    <Container fluid>
      <Row className="justify-content-md-center">
        <Col md="8">
          <Card className="login">
            <Card.Header>
              <Card.Title as="h4">Film keresése</Card.Title>
            </Card.Header>
            <Card.Body>
              <Form noValidate validated={isFormValidated}>
                <Row>
                  <Col md="12">
                    <Form.Group controlId="movie-query-field">
                      <Form.Label>Keresett film:</Form.Label>
                      <Form.Control
                        isInvalid={!!queryError}
                        onChange={onQueryChange}
                        required
                        value={queryText}
                      />
                      {feedback && (
                        <Form.Control.Feedback type="invalid">
                          {queryError}
                        </Form.Control.Feedback>
                      )}
                    </Form.Group>
                  </Col>
                </Row>
                <Button onClick={onSearch} variant="primary">
                  Keresés
                </Button>
              </Form>
              {searchError && <Alert variant="danger">{searchError}</Alert>}
            </Card.Body>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};

export default SearchForm;
