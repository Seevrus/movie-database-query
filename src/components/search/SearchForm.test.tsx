import {
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer, { act } from 'react-test-renderer';
import { MovieService } from '../../model/MovieService';
import movieServiceResponse from '../../test-data/movie-service-response';
import SearchForm from './SearchForm';

describe('SearchForm', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest
      .spyOn(MovieService, 'searchMovies')
      .mockResolvedValue(movieServiceResponse);
  });

  afterAll(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('should render correctly', () => {
    const component = renderer.create(
      <SearchForm
        movies={{
          query: '',
          pages: 42,
          movies: new Map(),
        }}
        setLoading={jest.fn()}
        setMovies={jest.fn()}
        activePage={1}
        setActivePage={jest.fn()}
        setNumberOfPages={jest.fn()}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should query for movies', async () => {
    const setActivePageMock = jest.fn();
    const setLoadingMock = jest.fn();
    const setMoviesMock = jest.fn();
    const setNumberOfPagesMock = jest.fn();
  
    render(
      <SearchForm
        movies={{
          query: '',
          pages: 42,
          movies: new Map(),
        }}
        setLoading={setLoadingMock}
        setMovies={setMoviesMock}
        activePage={1}
        setActivePage={setActivePageMock}
        setNumberOfPages={setNumberOfPagesMock}
      />,
    );

    act(() => {
      userEvent.type(
        screen.getByRole('textbox', { name: 'Keresett film:' }), 'star',
      );
    });

    jest.runAllTimers();

    await waitFor(() => {
      expect(setLoadingMock).toHaveBeenCalledTimes(2);
    });

    expect(setActivePageMock).toHaveBeenCalledWith(1);

    expect(setLoadingMock).toHaveBeenCalledTimes(2);
    expect(setLoadingMock).toHaveBeenNthCalledWith(1, true);
    expect(setLoadingMock).toHaveBeenNthCalledWith(2, false);

    expect(setMoviesMock).toHaveBeenCalledWith({
      query: 'star',
      pages: 1,
      movies: new Map().set(1, movieServiceResponse.movies),
    });

    expect(setNumberOfPagesMock).toHaveBeenCalledWith(1);
  });
});
