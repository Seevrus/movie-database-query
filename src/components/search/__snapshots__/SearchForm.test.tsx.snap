// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should render correctly 1`] = `
<div
  className="container-fluid"
>
  <div
    className="justify-content-md-center row"
  >
    <div
      className="col-md-8"
    >
      <div
        className="login card"
      >
        <div
          className="card-header"
        >
          <h4
            className="card-title"
          >
            Film keresése
          </h4>
        </div>
        <div
          className="card-body"
        >
          <form
            className=""
            noValidate={true}
          >
            <div
              className="row"
            >
              <div
                className="col-md-12"
              >
                <div>
                  <label
                    className="form-label"
                    htmlFor="movie-query-field"
                  >
                    Keresett film:
                  </label>
                  <input
                    className="form-control"
                    id="movie-query-field"
                    onChange={[Function]}
                    placeholder="Minimum három karakter..."
                    required={true}
                    value=""
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
